solar pv tracking option added

enrilizaga opened the pull request:
<!--
SPDX-FileCopyrightText: 2021 The Atlite Authors

SPDX-License-Identifier: CC0-1.0
-->

## Change proposed in this Pull Request

Added 1-axis vertical and 2-axis tracking option for solar pv (+ small documentation addition for get_windturbineconfig)

## Description
Updated orientation.py and convert.py:
- Default option is no tracking (tracking = 0)
- For vertical tracking, surface azimuth = sun_azimuth angles 
- For 2-axis tracking, both the azimuth and the slope follow the sun, thus cos(incidence angle)=1


## Motivation and Context
This change adds the solar pv tracking capability to the code. Solar pv tracking is widely used as it increases the solar pv efficiency significantly. This change improves the fidelity of the code to represent those efficiencies

## How Has This Been Tested?
The updated code was tested locally (see snapshots)
- The following commands provided the same results (tracking = 0 default)
cap_factors_pv_notrack = cutout.pv(panel='CSi', orientation={'slope': 30., 'azimuth': 180.}, capacity_factor=True)
cap_factors_pv_0axis = cutout.pv(panel='CSi', orientation={'slope': 30., 'azimuth': 180.}, tracking=0, capacity_factor=True)
![0tracking_comparison](https://user-images.githubusercontent.com/94380837/190222628-edfe6a58-cb14-4816-a305-0fa73fa7b403.PNG)
![0tracking_comparison_2](https://user-images.githubusercontent.com/94380837/190222623-05199507-16b0-4f4d-903b-07e5d8388a57.PNG)

- Tracking = 1 and =2 were tested as well:
cap_factors_pv_1axis = cutout.pv(panel='CSi', orientation={'slope': 30., 'azimuth': 180.}, tracking=1, capacity_factor=True)
cap_factors_pv_2axis = cutout.pv(panel='CSi', orientation={'slope': 30., 'azimuth': 180.}, tracking=2, capacity_factor=True)
![1tracking](https://user-images.githubusercontent.com/94380837/190222625-c50cfb6b-e6d8-41e5-84d5-cbb2ecc88b8a.PNG)
![2tracking](https://user-images.githubusercontent.com/94380837/190222627-60903bfb-6b6a-4849-95c6-1566b193dead.PNG)

- Checked that csp was not affected (tracking change only for solar pv)
cf = cutout.csp(installation="SAM_solar_tower", capacity_factor=True)
cf = cutout.csp(installation="SAM_parabolic_trough", capacity_factor=True)
![csp](https://user-images.githubusercontent.com/94380837/190223635-2bfaafe3-5ce4-4949-a778-189f0707686d.PNG)

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I tested my contribution locally and it seems to work fine.
- [x] I locally ran `pytest` inside the repository and no unexpected problems came up.
- [ ] I have adjusted the docstrings in the code appropriately.
- [ ] I have documented the effects of my code changes in the documentation `doc/`.
- [ ] I have added newly introduced dependencies to `environment.yaml` file.
- [ ] I have added a note to release notes `doc/release_notes.rst`.
- [ ] I have used `pre-commit run --all` to lint/format/check my contribution

FabianHofmann commented:
Hey @enrilizaga, sorry this takes a while on our side... but the code looks great! Some optional point would be awesome

* [ ] Could you add your notebook to the examples list? Then, we can also try it out, [see the docs on how to add it](https://atlite.readthedocs.io/en/latest/contributing.html#contributing-examples)
* [ ] A note in the release notes (`doc/release_notes.rst`) - just follow the pattern you see there.
* [ ] A test would be nice in `tests/test_preparation_and_conversion.py`

Would that be possible for you?

enrilizaga commented:
Hi @FabianHofmann, many thanks for your reply! I have: 

- [x] added notebook example solarpv_tracking_options.ipynb and corresponding .nblink

- [x] added pv_tracking_test function within test_preparation_and_conversion.py

- [x] added tracking option for trigon_model "simple" within tiltedirradiation function

- [x] added new version 0.2.10 in the RELEASE_NOTES.rst

Please let me know.
Thank you!

FabianHofmann commented:
This is awesome. I'm having a look at the changes next week.

fneum commented:
Ping! Would be super cool to include it in the upcoming PyPSA-Eur version.

fneum commented:
closes #140 
closes #220

FabianHofmann commented on atlite/convert.py:
```diff
@@ -501,7 +500,10 @@ def wind(cutout, turbine, smooth=False, **params):
         hub height and 'V', 'POW' defining the power curve.
         Alternatively a str refering to a local or remote turbine configuration
-        as accepted by atlite.resource.get_windturbineconfig().
+        as accepted by atlite.resource.get_windturbineconfig(). Locally stored turbine
+        configurations can also be modified with this function. E.g. to setup a different hub
+        height from the one used in the yaml file,one would write
+                "turbine=get_windturbineconfig(“NREL_ReferenceTurbine_5MW_offshore”)|{“hub_height”:120}"
```
I am wondering where this change is coming from.

fneum replied:
Good change!

enrilizaga replied:
@FabianHofmann just a small doc addition explaining how to modify the wind hub height. This is unrelated to the pv tracking update

FabianHofmann commented on atlite/convert.py:
```diff
@@ -530,21 +532,24 @@ def wind(cutout, turbine, smooth=False, **params):
 # solar PV
-def convert_pv(ds, panel, orientation, trigon_model="simple", clearsky_model="simple"):
+def convert_pv(
+    ds, panel, orientation, tracking, trigon_model="simple", clearsky_model="simple"
```
Can we make `tracking` a keyword argument here?

fneum replied:
Default to no tracking would be a sensible choice.

enrilizaga replied:
Just updated it. The no tracking option is the default option (e.g. see third assertion of pv_tracking_test)

FabianHofmann commented on atlite/convert.py:
```diff
@@ -561,6 +566,10 @@ def pv(cutout, panel, orientation, clearsky_model=None, **params):
         'azimuth': 0.0} or a callback function with the same signature
         as the callbacks generated by the
         'atlite.pv.orientation.make_*' functions.
+    tracking : int 0, 1, 2:
+        0 for no tracking, default
+        1 for 1-axis vertical tracking
+        2 for 2-axis tracking
```
I am wondering whether we should rather take the arguments 
* False, 
* 'v' or 'vertical' 
* 'h' or 'horizontal' 
* 'full'

where horizontal is not supported right now. This would make integration in the future easier. Could you change this?

fneum replied:
or instead of full "hv" or "vh"

enrilizaga replied:
good idea indeed! I have not included 'horizontal' as that is not supported by the code, but should be now easier to integrate it in the future

martavp replied:
Sorry to jump in here without actually reviewing the code but this comment caught my attention. I think we should prioritize including 1-axis horizontal tracking. 
In short, horizontal 1-axis tracking is widely used in the market today (about 50% of new PV power plants use it) while 2-axis or vertical tracking is not used at all. 
I understand that this PR only includes vertical and 2-axes (and that is great, thanks for the work @enrilizaga ) but I think it would look weird in the next release if only exotic options are included by not the dominant "horizontal 1-axis tracking".

enrilizaga replied:
Thanks @martavp. I understand your point, but I am not sure why that would prevent this PR from being added into the code? Horizontal 1-axis tracking may be more widely used, but it is also more complex to implement and outside my scope of work. PS: would be great if you could share the reference re PV panels market share by tracking type!

FabianHofmann replied:
@enrilizaga no worries, the PR is great (!) and this missing feature would not prevent it from being merged. But since you already have experience with the math, would you mind lifting the implementation to the full features stack? That would make it perfect. I had a try with horizontal tracking in #220 but due to missing time, I did not dig down any further. There you also find a reference (https://www.nrel.gov/docs/fy13osti/58891.pdf) for the hsat.

enrilizaga replied:
Got it, thanks for the clarification. When is next release? I could look into it at some point in the next month or so but I am currently time constrained.

Re the market share, thanks for sharing that reference. Unfortunately ITRPV does not seem to have data on horizontal vs vertical 1-axis tracking separately and just reports all as "1-axis tracking"?

FabianHofmann replied:
No pressure on the release date. If you can make it this year, that would be great. Otherwise I could also spend some hours, likely not before January on this. To give you a proper hint on the implementation I'd have make some research, which is atm not possible for me...

martavp replied:
Hi @enrilizaga I'm pretty sure that ITRPV only mentions "1-axis tracking" assuming that is horizontal tracking. Vertical tracking is not used at all in commercial utility-scale applications. 

I'm also time constrained until January, but I can also find some time there to make suggestions. I have also talked to @Parisra and she might be able to put some time into implementing this.

FabianHofmann reviewed (changes requested):
@enrilizaga the code look great! Some comments from my side. I am currently running the notebook. I also want to make a check that the pv for tracking=0 is the same as in the current master.

enrilizaga commented:
@FabianHofmann @fneum thanks for your comments. Just included them in the last commit. Please let me know!

FabianHofmann commented:
Hey @enrilizaga, do you have concrete plans on continuing on this PR? Otherwise, we could look whether we can find someone to finish it up.

martavp commented:
I know that @Parisra is working on implementing the horizontal tracking option. 
Mainly obtaining the tilt and orientation of the solar panels using Eq. 8-10 from [NREL report](https://www.nrel.gov/docs/fy13osti/58891.pdf)
Maybe @FabianHofmann you can accept @enrilizaga PR and @Parisra takes it from there

fneum commented:
I think that’d make sense. No need for full tracking implemented until merging.

FabianHofmann commented:
@martavp let's do it :) @Parisra could you then make sure to counter test the formulas? I haven't had the time to go into them in detail. 

@enrilizaga many thanks again! This is a great contribution and super well done!

Parisra commented:
Hi @FabianHofmann , yes I can do that. I've just finished adding the formulas for horizontal tracking and will be doing some tests next week to make sure the results make sense. I'll do a pull request once everything is ready.

FabianHofmann commented:
great thanks @Parisra
